ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/streaming/BedrockStreamingHandler.java

```diff
@@ -0,0 +1,435 @@
+            AtomicReference<Map<String, Object>> toolInput = new AtomicReference<>();
+            AtomicReference<String> toolUseId = new AtomicReference<>();
+            StringBuilder toolInputAccumulator = new StringBuilder();
+
+            // Build Bedrock client
+            BedrockRuntimeAsyncClient bedrockClient = buildBedrockRuntimeAsyncClient();
+
+            // Parse payload to build ConverseStreamRequest
+            ConverseStreamRequest request = buildConverseStreamRequest(payload, parameters);
+
+            // Build response handler with comprehensive raw logging
+            ConverseStreamResponseHandler handler = ConverseStreamResponseHandler.builder().onResponse(response -> {}).onError(error -> {
+                log.error("Converse stream error: {}", error.getMessage());
+                listener.onFailure(new MLException("Error from remote service: " + error.getMessage(), error));
+            }).onComplete(() -> {
+                if (!functionCallInProgress.get() && !agentExecutionInProgress.get()) {
+                    sendCompletionResponse(isStreamClosed, listener);
+                } else {
+                    log.debug("Function call or agent execution in progress - keeping stream open");
+                }
+            }).subscriber(event -> {
+                handleStreamEvent(
+                    event,
+                    listener,
+                    isStreamClosed,
+                    functionCallInProgress,
+                    agentExecutionInProgress,
+                    toolName,
+                    toolInput,
+                    toolUseId,
+                    toolInputAccumulator
+                );
+            }).build();
+
+            // Start streaming
+            bedrockClient.converseStream(request, handler);
+        } catch (Exception e) {
+            log.error("Failed to execute Bedrock streaming", e);
+            handleError(e, listener);
+        }
+    }
+
+    private ConverseStreamRequest buildConverseStreamRequest(String payload, Map<String, String> parameters) {
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode payloadJson = mapper.readTree(payload);
+            ConverseStreamRequest.Builder requestBuilder = ConverseStreamRequest.builder().modelId(parameters.get("model"));
+
+            // Add system messages if present
+            if (payloadJson.has("system")) {
+                requestBuilder.system(parseSystemMessages(payloadJson.get("system")));
+            }
+
+            // Add messages if present
+            if (payloadJson.has("messages")) {
+                requestBuilder.messages(parseMessages(payloadJson.get("messages")));
+            }
+
+            // Add tool configuration if present
+            if (payloadJson.has("toolConfig")) {
+                requestBuilder.toolConfig(parseToolConfig(payloadJson.get("toolConfig")));
+            }
+            return requestBuilder.build();
```

Comment:
Why not use the ConverseStreamRequest.builder() to simplify this block? 
return ConverseStreamRequest.builder()
            .modelId(parameters.get("model"))
            .system(getOptionalNode(payloadJson, "system").map(this::parseSystemMessages).orElse(null))
            .messages(getOptionalNode(payloadJson, "messages").map(this::parseMessages).orElse(null))
            .toolConfig(getOptionalNode(payloadJson, "toolConfig").map(this::parseToolConfig).orElse(null))
            .build();

private Optional<JsonNode> getOptionalNode(JsonNode json, String field) {
    return Optional.ofNullable(json.get(field));
}